File: src/apps/daemon/internal/networking/tlsdecoy/multi_strategy_fragmenter_test.go
package tlsdecoy

import (
	"bytes"
	"testing"
	"time"
)

func buildTestClientHello(sniHost string) []byte {
	hostBytes := []byte(sniHost)
	var pkt []byte

	// TLS Record Header: Handshake (0x16), TLS 1.0 (0x03, 0x01), Length placeholder
	pkt = append(pkt, 0x16, 0x03, 0x01, 0x00, 0x00)

	handshakeStart := len(pkt)
	// Handshake Header: ClientHello (0x01), Handshake Length placeholder (3 bytes)
	pkt = append(pkt, 0x01, 0x00, 0x00, 0x00)

	// Version: TLS 1.2 (0x03, 0x03)
	pkt = append(pkt, 0x03, 0x03)

	// Random: 32 bytes
	pkt = append(pkt, bytes.Repeat([]byte{0x42}, 32)...)

	// Session ID: len 0
	pkt = append(pkt, 0x00)

	// Cipher Suites: 2 bytes len + 2 suites (4 bytes)
	pkt = append(pkt, 0x00, 0x04, 0x13, 0x01, 0x13, 0x02)

	// Compression: len 1 + null compression (0x00)
	pkt = append(pkt, 0x01, 0x00)

	// Extensions
	extLenPos := len(pkt)
	pkt = append(pkt, 0x00, 0x00) // placeholder

	extStart := len(pkt)

	// Extension: Server Name Indication (0x0000)
	sniExtLen := 2 + 1 + 2 + len(hostBytes)
	pkt = append(pkt, 0x00, 0x00, byte((sniExtLen>>8)&0xff), byte(sniExtLen&0xff))

	listLen := 1 + 2 + len(hostBytes)
	pkt = append(pkt, byte((listLen>>8)&0xff), byte(listLen&0xff), 0x00, byte((len(hostBytes)>>8)&0xff), byte(len(hostBytes)&0xff))
	pkt = append(pkt, hostBytes...)

	// Fill extensions length
	extLen := len(pkt) - extStart
	pkt[extLenPos] = byte((extLen >> 8) & 0xff)
	pkt[extLenPos+1] = byte(extLen & 0xff)

	// Fill handshake length
	hsLen := len(pkt) - handshakeStart - 4
	pkt[handshakeStart+1] = byte((hsLen >> 16) & 0xff)
	pkt[handshakeStart+2] = byte((hsLen >> 8) & 0xff)
	pkt[handshakeStart+3] = byte(hsLen & 0xff)

	// Fill record length
	recLen := len(pkt) - 5
	pkt[3] = byte((recLen >> 8) & 0xff)
	pkt[4] = byte(recLen & 0xff)

	return pkt
}

func TestLocateSNI_And_ExtractSNI(t *testing.T) {
	pkt := buildTestClientHello("speedtest.net")
	offset, length, found := LocateSNI(pkt)
	if !found {
		t.Fatalf("LocateSNI failed to find SNI")
	}
	if string(pkt[offset:offset+length]) != "speedtest.net" {
		t.Fatalf("unexpected extracted host slice: %s", string(pkt[offset:offset+length]))
	}

	extracted := ExtractSNI(pkt)
	if extracted != "speedtest.net" {
		t.Fatalf("unexpected ExtractSNI output: %s", extracted)
	}
}

func TestTenFragmentationStrategies(t *testing.T) {
	pkt := buildTestClientHello("speedtest.net")
	settings := DefaultFinalMaskSettings()

	// 1. Raw
	raw := FragmentPacket(pkt, FragmentStrategyRaw, settings)
	if len(raw) != 1 || !bytes.Equal(raw[0], pkt) {
		t.Fatalf("Raw fragmentation failed")
	}

	// 2. Full5
	full5 := FragmentPacket(pkt, FragmentStrategyFull5, settings)
	if len(full5) <= 1 || len(full5[0]) != 5 {
		t.Fatalf("Full5 chunking failed")
	}
	if !bytes.Equal(bytes.Join(full5, nil), pkt) {
		t.Fatalf("Full5 reconstruction failed")
	}

	// 3. Full10
	full10 := FragmentPacket(pkt, FragmentStrategyFull10, settings)
	if len(full10[0]) != 10 || !bytes.Equal(bytes.Join(full10, nil), pkt) {
		t.Fatalf("Full10 chunking failed")
	}

	// 4. Full20
	full20 := FragmentPacket(pkt, FragmentStrategyFull20, settings)
	if len(full20[0]) != 20 || !bytes.Equal(bytes.Join(full20, nil), pkt) {
		t.Fatalf("Full20 chunking failed")
	}

	// 5. Half
	half := FragmentPacket(pkt, FragmentStrategyHalf, settings)
	if len(half) != 2 || !bytes.Equal(bytes.Join(half, nil), pkt) {
		t.Fatalf("Half fragmentation failed")
	}

	// 6. SniBoundary
	sniBoundary := FragmentPacket(pkt, FragmentStrategySniBoundary, settings)
	if len(sniBoundary) != 2 || !bytes.Equal(bytes.Join(sniBoundary, nil), pkt) {
		t.Fatalf("SniBoundary fragmentation failed")
	}
	offset, _, _ := LocateSNI(pkt)
	if len(sniBoundary[0]) != offset {
		t.Fatalf("SniBoundary expected offset %d, got %d", offset, len(sniBoundary[0]))
	}

	// 7. SniSplit
	sniSplit := FragmentPacket(pkt, FragmentStrategySniSplit, settings)
	if len(sniSplit) != 2 || !bytes.Equal(bytes.Join(sniSplit, nil), pkt) {
		t.Fatalf("SniSplit fragmentation failed")
	}

	// 8. TlsRecordFrag
	tlsFrag := FragmentPacket(pkt, FragmentStrategyTlsRecordFrag, settings)
	if len(tlsFrag) != 2 || tlsFrag[0][0] != 0x16 || tlsFrag[1][0] != 0x16 {
		t.Fatalf("TlsRecordFrag failed")
	}

	// 9. TlsSniRecords
	tlsSni := FragmentPacket(pkt, FragmentStrategyTlsSniRecords, settings)
	if len(tlsSni) != 2 || tlsSni[0][0] != 0x16 || tlsSni[1][0] != 0x16 {
		t.Fatalf("TlsSniRecords failed")
	}

	// 10. FinalMaskTlsHello
	finalMask := FragmentPacket(pkt, FragmentStrategyFinalMaskTlsHello, settings)
	if len(finalMask) == 0 {
		t.Fatalf("FinalMaskTlsHello failed")
	}
}

func TestRewriteFinalMaskWrites(t *testing.T) {
	pkt := buildTestClientHello("ignitelimit.com")
	settings := DefaultFinalMaskSettings()

	rewrite := RewriteFinalMaskWrites(pkt, settings)
	if len(rewrite.FirstWrite) < 10 || rewrite.FirstWrite[0] != 0x16 {
		t.Fatalf("RewriteFinalMaskWrites first write invalid")
	}
	recordLen := (int(rewrite.FirstWrite[3]) << 8) | int(rewrite.FirstWrite[4])
	if recordLen != 5 {
		t.Fatalf("expected record len 5, got %d", recordLen)
	}

	contiguous := RewriteFinalMaskTlsHello(pkt, settings)
	if !bytes.Equal(contiguous, rewrite.FirstWrite) {
		t.Fatalf("contiguous bytes mismatch")
	}
}

func TestCarrierRouteSelector(t *testing.T) {
	edge1 := CarrierEdgeProfile{Address: "104.18.1.1", Port: 443, Role: "primary", FinalmaskMaxSplit: 2}
	edge2 := CarrierEdgeProfile{Address: "104.18.1.2", Port: 443, Role: "irancell", FinalmaskMaxSplit: 100}
	edge3 := CarrierEdgeProfile{Address: "172.66.0.1", Port: 443, Role: "fallback", FinalmaskMaxSplit: 2}

	edges := []CarrierEdgeProfile{edge1, edge2, edge3}
	selector := NewCarrierRouteSelector(100 * time.Millisecond)

	ordered := selector.OrderedEdges(edges)
	if len(ordered) != 3 || ordered[0].Address != "104.18.1.1" {
		t.Fatalf("unexpected ordered edges initial: %v", ordered)
	}

	selector.RecordFailure(edge1)
	if !selector.IsInCooldown(edge1) {
		t.Fatalf("edge1 should be in cooldown")
	}

	orderedAfterFail := selector.OrderedEdges(edges)
	if orderedAfterFail[0].Address != "104.18.1.2" || orderedAfterFail[2].Address != "104.18.1.1" {
		t.Fatalf("edge1 should be sorted to the end: %v", orderedAfterFail)
	}

	selector.RecordSuccess(edge1)
	if selector.IsInCooldown(edge1) {
		t.Fatalf("edge1 cooldown should be cleared")
	}
	orderedAfterSuccess := selector.OrderedEdges(edges)
	if orderedAfterSuccess[0].Address != "104.18.1.1" {
		t.Fatalf("edge1 should be restored to front: %v", orderedAfterSuccess)
	}
}

File: src/packages/lumicore/tests/multi_strategy_fragmenter_test.rs
use lumicore::evasion::multi_strategy_fragmenter::{
    extract_sni, fixed_chunks, fragment_packet, locate_sni, rewrite_final_mask_tls_hello,
    rewrite_final_mask_writes, split_at, split_tls_record, CarrierEdgeProfile,
    CarrierRouteSelector, FinalMaskSettings, FragmentStrategy,
};
use std::time::Duration;

/// Helper to build a minimal valid TLS 1.2/1.3 ClientHello with SNI extension.
fn build_test_client_hello(sni_host: &str) -> Vec<u8> {
    let host_bytes = sni_host.as_bytes();
    let mut pkt = Vec::new();

    // TLS Record Header: Handshake (0x16), TLS 1.0 (0x03, 0x01), Length placeholder
    pkt.extend_from_slice(&[0x16, 0x03, 0x01, 0x00, 0x00]);

    let handshake_start = pkt.len();
    // Handshake Header: ClientHello (0x01), Handshake Length placeholder (3 bytes)
    pkt.extend_from_slice(&[0x01, 0x00, 0x00, 0x00]);

    // Version: TLS 1.2 (0x03, 0x03)
    pkt.extend_from_slice(&[0x03, 0x03]);

    // Random: 32 bytes
    pkt.extend_from_slice(&[0x42; 32]);

    // Session ID: len 0
    pkt.push(0x00);

    // Cipher Suites: 2 bytes len + 2 suites (4 bytes)
    pkt.extend_from_slice(&[0x00, 0x04, 0x13, 0x01, 0x13, 0x02]);

    // Compression: len 1 + null compression (0x00)
    pkt.extend_from_slice(&[0x01, 0x00]);

    // Extensions
    let ext_len_pos = pkt.len();
    pkt.extend_from_slice(&[0x00, 0x00]); // placeholder

    let ext_start = pkt.len();

    // Extension: Server Name Indication (0x0000)
    let sni_ext_len = 2 + 1 + 2 + host_bytes.len(); // list_len(2) + name_type(1) + name_len(2) + host
    pkt.extend_from_slice(&[
        0x00, 0x00,
        ((sni_ext_len >> 8) & 0xff) as u8,
        (sni_ext_len & 0xff) as u8,
    ]);

    let list_len = 1 + 2 + host_bytes.len();
    pkt.extend_from_slice(&[
        ((list_len >> 8) & 0xff) as u8,
        (list_len & 0xff) as u8,
        0x00, // host_name type
        ((host_bytes.len() >> 8) & 0xff) as u8,
        (host_bytes.len() & 0xff) as u8,
    ]);
    pkt.extend_from_slice(host_bytes);

    // Fill extensions length
    let ext_len = pkt.len() - ext_start;
    pkt[ext_len_pos] = ((ext_len >> 8) & 0xff) as u8;
    pkt[ext_len_pos + 1] = (ext_len & 0xff) as u8;

    // Fill handshake length
    let hs_len = pkt.len() - handshake_start - 4;
    pkt[handshake_start + 1] = ((hs_len >> 16) & 0xff) as u8;
    pkt[handshake_start + 2] = ((hs_len >> 8) & 0xff) as u8;
    pkt[handshake_start + 3] = (hs_len & 0xff) as u8;

    // Fill record length
    let rec_len = pkt.len() - 5;
    pkt[3] = ((rec_len >> 8) & 0xff) as u8;
    pkt[4] = (rec_len & 0xff) as u8;

    pkt
}

#[test]
fn test_sni_locating_and_extraction() {
    let client_hello = build_test_client_hello("speedtest.net");
    let located = locate_sni(&client_hello);
    assert!(located.is_some(), "Must locate SNI in valid ClientHello");

    let (offset, len) = located.unwrap();
    assert_eq!(&client_hello[offset..offset + len], b"speedtest.net");

    let extracted = extract_sni(&client_hello);
    assert_eq!(extracted, Some("speedtest.net".to_string()));
}

#[test]
fn test_ten_fragmentation_strategies() {
    let client_hello = build_test_client_hello("speedtest.net");
    let settings = FinalMaskSettings::default();

    // 1. Raw
    let raw = fragment_packet(&client_hello, FragmentStrategy::Raw, &settings);
    assert_eq!(raw.len(), 1);
    assert_eq!(raw[0], client_hello);

    // 2. Full5
    let full5 = fragment_packet(&client_hello, FragmentStrategy::Full5, &settings);
    assert!(full5.len() > 1);
    assert_eq!(full5[0].len(), 5);
    let reconstructed: Vec<u8> = full5.into_iter().flatten().collect();
    assert_eq!(reconstructed, client_hello);

    // 3. Full10
    let full10 = fragment_packet(&client_hello, FragmentStrategy::Full10, &settings);
    assert_eq!(full10[0].len(), 10);
    let reconstructed10: Vec<u8> = full10.into_iter().flatten().collect();
    assert_eq!(reconstructed10, client_hello);

    // 4. Full20
    let full20 = fragment_packet(&client_hello, FragmentStrategy::Full20, &settings);
    assert_eq!(full20[0].len(), 20);
    let reconstructed20: Vec<u8> = full20.into_iter().flatten().collect();
    assert_eq!(reconstructed20, client_hello);

    // 5. Half
    let half = fragment_packet(&client_hello, FragmentStrategy::Half, &settings);
    assert_eq!(half.len(), 2);
    let reconstructed_half: Vec<u8> = half.into_iter().flatten().collect();
    assert_eq!(reconstructed_half, client_hello);

    // 6. SniBoundary
    let sni_boundary = fragment_packet(&client_hello, FragmentStrategy::SniBoundary, &settings);
    assert_eq!(sni_boundary.len(), 2);
    let (sni_offset, _) = locate_sni(&client_hello).unwrap();
    assert_eq!(sni_boundary[0].len(), sni_offset);
    let reconstructed_boundary: Vec<u8> = sni_boundary.into_iter().flatten().collect();
    assert_eq!(reconstructed_boundary, client_hello);

    // 7. SniSplit
    let sni_split = fragment_packet(&client_hello, FragmentStrategy::SniSplit, &settings);
    assert_eq!(sni_split.len(), 2);
    let reconstructed_split: Vec<u8> = sni_split.into_iter().flatten().collect();
    assert_eq!(reconstructed_split, client_hello);

    // 8. TlsRecordFrag
    let tls_frag = fragment_packet(&client_hello, FragmentStrategy::TlsRecordFrag, &settings);
    assert_eq!(tls_frag.len(), 2);
    assert_eq!(tls_frag[0][0], 0x16);
    assert_eq!(tls_frag[1][0], 0x16);

    // 9. TlsSniRecords
    let tls_sni_records = fragment_packet(&client_hello, FragmentStrategy::TlsSniRecords, &settings);
    assert_eq!(tls_sni_records.len(), 2);
    assert_eq!(tls_sni_records[0][0], 0x16);
    assert_eq!(tls_sni_records[1][0], 0x16);

    // 10. FinalMaskTlsHello
    let final_mask = fragment_packet(&client_hello, FragmentStrategy::FinalMaskTlsHello, &settings);
    assert!(!final_mask.is_empty());
}

#[test]
fn test_final_mask_rewriting() {
    let client_hello = build_test_client_hello("ignitelimit.com");
    let settings = FinalMaskSettings {
        packet: "tlshello".to_string(),
        length: 5,
        delay_ms: 0,
        max_split: 2,
    };

    let rewrite = rewrite_final_mask_writes(&client_hello, &settings);
    assert!(rewrite.first_write.len() > 10);
    assert_eq!(rewrite.first_write[0], 0x16); // First TLS record
    assert_eq!(((rewrite.first_write[3] as usize) << 8) | (rewrite.first_write[4] as usize), 5); // 5 bytes payload in record 1

    let contiguous = rewrite_final_mask_tls_hello(&client_hello, &settings);
    assert_eq!(contiguous, rewrite.first_write);
}

#[test]
fn test_carrier_route_selector_cooldown() {
    let edge1 = CarrierEdgeProfile::new("104.18.1.1", 443, "primary", 2);
    let edge2 = CarrierEdgeProfile::new("104.18.1.2", 443, "irancell", 100);
    let edge3 = CarrierEdgeProfile::new("172.66.0.1", 443, "fallback", 2);

    let edges = vec![edge1.clone(), edge2.clone(), edge3.clone()];
    let mut selector = CarrierRouteSelector::new(Duration::from_millis(100));

    // Initially all are healthy
    let ordered = selector.ordered_edges(&edges);
    assert_eq!(ordered, edges);

    // Mark edge1 as failed
    selector.record_failure(&edge1);
    assert!(selector.is_in_cooldown(&edge1));

    // Ordered list should place edge1 at the end
    let ordered_after_fail = selector.ordered_edges(&edges);
    assert_eq!(ordered_after_fail[0], edge2);
    assert_eq!(ordered_after_fail[1], edge3);
    assert_eq!(ordered_after_fail[2], edge1);

    // Record success on edge1 clears cooldown immediately
    selector.record_success(&edge1);
    assert!(!selector.is_in_cooldown(&edge1));
    let ordered_after_success = selector.ordered_edges(&edges);
    assert_eq!(ordered_after_success[0], edge1);
}
